GDPR message banner, refs #12180

sbreker opened the pull request:
This PR includes work to add the GDPR banner. The GDPR banner can be dismissed with the OK button - when this is performed, a session attribute is created that will hide the banner for the session. This works for both authenticated and unauthenticated visitors. The banner functionality and text have a new settings panel under Admin->Settings called "Privacy Notification" - here the GDPR banner functionality can be deactivated, and the message can be customized. The GDPR will be ON by default. Also added a new privacy message static page which is linked from the Quick Links menu as well as from the notification banner.

sbreker committed:
Add Privacy Notification config GDPR, refs #12180

Added Settings page and tied into settings sidebar menu.
Added Privacy statement static page setting.

sbreker committed:
Add GDPR Quick Links menu item, refs #12180

Added Quick Links menu item 'Privacy Policy'.

sbreker committed:
Add privacy message component, refs #12180

Add privacyMessage component to header template.

sbreker committed:
Hide the GDPR component in print.css, refs #12180

Don't print the GDPR message component.

sbreker committed:
Add GDPR javascript, refs #12180

Javascript file that hides the GDPR message once OK button is clicked.

sbreker committed:
GDPR component action and template, refs #12180

Add GDPR privacy message component action and template.
Add GDPR controller that sets user session attribute when called from
js/privacyMessage.js.

sbreker committed:
GDPR message CSS changes, refs #12180

CSS changes to support GDPR banner message layout/formatting.

sbreker committed:
Default privacy static page content, refs #12180

Updated static page default content.

jraddaoui commented on apps/qubit/modules/default/actions/privacyMessageComponent.class.php:
```diff
@@ -0,0 +1,41 @@
+ */
+
+/**
+ * Check for updates component
```
;)

sbreker replied:
Fixed.

jraddaoui commented on apps/qubit/modules/default/actions/privacyMessageComponent.class.php:
```diff
@@ -0,0 +1,41 @@
+{
+  public function execute($request)
+  {
+    if ((null !== $this->privacyMessageDismissed = $this->context->user->getAttribute('privacy_message_dismissed')) ||
```
No need to save it in `$this->privacyMessageDismissed``, right?

sbreker replied:
Correct. Removed.

jraddaoui commented on apps/qubit/modules/default/actions/privacyMessageComponent.class.php:
```diff
@@ -0,0 +1,41 @@
+    {
+      return sfView::NONE;
+    }
+    else
```
With the `return` in the previous `if` this `else` is not needed and adds unnecessary indentation.

jraddaoui commented on apps/qubit/modules/default/templates/_privacyMessage.php:
```diff
@@ -0,0 +1,12 @@
+<script type="text/javascript" charset="utf-8" src="/js/privacyMessage.js"></script>
```
I think it's better to add it in the component, so it gets added to the head section instead if in the body. E.g.:

https://github.com/artefactual/atom/blob/qa%2F2.5.x/apps/qubit/modules/informationobject/actions/indexAction.class.php#L122

sbreker replied:
I have tried to make this work by adding the javascript in the controller but it has no effect even after clearing cache, restarting services etc. I believe that this does not work for classes extending sfComponent. I have tried including as in your example which extends sfAction, but it has no effect - the js is not included in the rendered page. I can find other classes in AtoM which extend sfComponent that have also included the js in the template.

sbreker replied:
I have also tried loading this using the view.yml config file for default module, but this has the side effect of loading the js on every AtoM page access, even when it is not needed. Placing this in the template was the only way I could find to include this js successfully, but only when the banner is actually displayed and the js is required.

jraddaoui replied:
I didn't know that, thanks!

jraddaoui commented on data/fixtures/settings.yml:
```diff
@@ -878,6 +878,16 @@ QubitSetting:
+    deleteable: 0
+    source_culture: en
+    value:
+      en: 'This website uses cookies to enhance your ability to browse and load content. [More Info.](/privacy)'
```
Nice use of the new Markdown feature. Being an editable setting I think it's okay to use it.

sbreker replied:
👍

jraddaoui commented on apps/qubit/modules/default/actions/privacyMessageComponent.class.php:
```diff
@@ -0,0 +1,41 @@
+    }
+    else
+    {
+      $this->notificationMessage = sfConfig::get('app_privacy_notification');
```
I'm not sure how is this going to work with the settings cache and translations. I think it's better to use the QubitSetting value in this case.

jraddaoui commented on lib/task/migrate/migrations/arMigration0165.class.php:
```diff
@@ -0,0 +1,98 @@
+ */
+
+/*
+ * Add taxonomy and terms for user actions
```
;)

sbreker replied:
Fixed.

jraddaoui commented on plugins/arDominionPlugin/css/less/print.less:
```diff
@@ -3,6 +3,7 @@
   /* Hide */
 
   #update-check,
+  #privacy-message,
```
Good catch.

jraddaoui commented on plugins/arDominionPlugin/css/less/scaffolding.less:
```diff
@@ -1552,6 +1552,37 @@ body.login #content {
+  margin-bottom: 0px;
+  margin-top: 0px;
+  text-align: center;
+  padding: 8px 35px 18px 14px !important;
```
Is this `!important` really necessary, I'd try to find a better selector to see if we can avoid using it.

sbreker replied:
Testing this again, it looks like '!important' is not necessary after all. Removed....

jraddaoui commented on plugins/arDominionPlugin/css/less/scaffolding.less:
```diff
@@ -1552,6 +1552,37 @@ body.login #content {
+      color: @gray;
+    }
+    a:hover {
+      color: darken(@linkColorHover, 10%);
```
Nice! It will help in custom themes.

jraddaoui reviewed (changes requested):
Nicely done @sbreker!

A few comments but nothing big.

sbreker committed:
Address CR feedback, refs #12180

Addressed feedback from code review. Only change not made is moving js
to controller. I believe for an sfComponent this must be included in
either the template, or the view.yml.

jraddaoui reviewed (approved):
Great! Thanks Steve!
